### PR TITLE
overlord/snapstate: setup/remove aliases as we link/unlink snaps

### PR DESCRIPTION
**Pull request by pedronis**
This also starts the infrastructure to track aliases' state in a "aliases"=>AliasState map.

setup-aliases considers only Enabled for the time being. This will change when we will implement "unalias".

### CODE REVIEW

**Comment by pedronis** [comment]
based on #2418

File: overlord/snapstate/aliases.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+func checkSnapAliasConflict(st *state.State, snapName string) error {
+	aliasStates, err := getAliases(st)
+	if err != nil {
+		return err
+	}
+	prefix := fmt.Sprintf("%s.", snapName)
+	for alias, aliasState := range aliasStates {
+		if alias == snapName || strings.HasPrefix(alias, prefix) {
+			if aliasState.Enabled != "" {
+				return fmt.Errorf("snap %q command namespace conflicts with enabled alias %q for %q", snapName, alias, aliasState.Enabled)
```
Comment (niemeyer) [inline overlord/snapstate/aliases.go]: Yeah, something like `Snap` is definitely a better name for `Enabled`.

File: overlord/snapstate/aliases.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+			aliases = append(aliases, &backend.Alias{
+				Name:   alias,
+				Target: filepath.Base(aliasApp.WrapperPath()),
+			})
+		}
+		// TODO: check Disabled => remove
+	}
+	aliases, err = m.backend.MissingAliases(aliases)
+	if err != nil {
+		return fmt.Errorf("cannot establish missing enabled aliases for snap %q: %v", snapName, err)
```
Comment (niemeyer) [inline overlord/snapstate/aliases.go]: Maybe `"cannot list aliases for %q snap: %v"` here and below

File: overlord/snapstate/aliases.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+				continue
+			}
+			aliases = append(aliases, &backend.Alias{
+				Name:   alias,
+				Target: filepath.Base(aliasApp.WrapperPath()),
+			})
+		}
+		// TODO: check Disabled => remove
+	}
+	aliases, err = m.backend.MissingAliases(aliases)
```
Comment (niemeyer) [inline overlord/snapstate/aliases.go]: Should we perhaps check these aliases?  Having non-matching (!= target) aliases here would be an inconsistency, and it'd never get fixed due to this.

File: overlord/snapstate/snapmgr_test.go [hunk 4]
```diff
@@ -870,7 +893,10 @@ func (s *snapmgrTestSuite) TestInstallRunThrough(c *C) {
 			name:  "some-snap",
 			revno: snap.R(42),
 		},
-	})
+	}
+	// start with an easier-to-read error if this fails:
```
Comment (chipaca) [inline overlord/snapstate/snapmgr_test.go]: :-)

**Pull request merged**